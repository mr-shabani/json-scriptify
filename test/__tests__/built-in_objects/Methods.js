/* eslint-disable no-undef */
var scriptify = require("../../../");
var checkSimilarity = require("../../object_similarity");

var run = function(obj, withFunctions) {
	if (withFunctions) return eval(scriptify.withAllFunctions(obj));
	return eval(scriptify(obj));
};

var checkFor = function(obj, withFunctions) {
	if (withFunctions) return checkSimilarity(obj, run(obj, withFunctions));
	return checkSimilarity(obj, run(obj));
};

test("simple method", function() {
	var o = {
		func(x) {
			return x;
		}
	};
	expect(scriptify(o)).toBe("({func(x) {\n      return x;\n    }})");
	var obj = o.func;
	obj.x = 1;
	obj.f = obj;
	expect(checkFor(obj)).toEqual(true);
	expect(checkFor({ f: obj })).toEqual(true);
	o = {
		function(x) {
			return x;
		}
	};
	obj = o.function;
	obj.x = 1;
	obj.f = obj;
	expect(checkFor(obj)).toEqual(true);
	expect(checkFor({ f: obj })).toEqual(true);
	o = {
		func: function func(x) {
			return x;
		}
	};
	obj = o.func;
	obj.x = 1;
	obj.f = obj;
	expect(checkFor(obj)).toEqual(true);
	expect(checkFor({ f: obj })).toEqual(true);
});

test("getter method", function() {
	var o = {
		get func() {
			return 1;
		}
	};
	expect(scriptify(o)).toBe("({get func() {\n      return 1;\n    }})");
	var obj = Object.getOwnPropertyDescriptor(o, "func").get;
	obj.x = 1;
	obj.f = obj;
	expect(checkFor(obj)).toEqual(true);
	expect(checkFor({ f: obj })).toEqual(true);
	obj = {};
	Object.defineProperty(obj, "prop", {
		get: Object.getOwnPropertyDescriptor(o, "func").get,
		configurable: true
	});
	obj.x = 1;
	obj.f = obj;
	expect(checkFor(obj)).toEqual(true);
	expect(checkFor({ f: obj })).toEqual(true);
});

test("setter method", function() {
	var o = {
		set func(x) {}
	};
	expect(scriptify(o)).toBe("({set func(x) {}})");
	var obj = Object.getOwnPropertyDescriptor(o, "func").set;
	obj.x = 1;
	obj.f = obj;
	expect(checkFor(obj)).toEqual(true);
	expect(checkFor({ f: obj })).toEqual(true);
	obj = {};
	Object.defineProperty(obj, "prop", {
		set: Object.getOwnPropertyDescriptor(o, "func").set,
		configurable: true
	});
	obj.x = 1;
	obj.f = obj;
	expect(checkFor(obj)).toEqual(true);
	expect(checkFor({ f: obj })).toEqual(true);
});

test("Generator methods", function() {
	var o = {
		*func(x) {
			yield x;
		}
	};
	expect(scriptify(o)).toBe("({*func(x) {\n      yield x;\n    }})");
	var obj = o.func;
	obj.x = 1;
	obj.f = obj;
	expect(checkFor(obj)).toEqual(true);
	expect(checkFor({ f: obj })).toEqual(true);
	o = {
		*function(x) {
			yield x;
		}
	};
	obj = o.function;
	obj.x = 1;
	obj.f = obj;
	expect(checkFor(obj)).toEqual(true);
	expect(checkFor({ f: obj })).toEqual(true);
	o = {
		func: function* func(x) {
			yield x;
		}
	};
	obj = o.func;
	obj.x = 1;
	obj.f = obj;
	expect(checkFor(obj)).toEqual(true);
	expect(checkFor({ f: obj })).toEqual(true);
});
