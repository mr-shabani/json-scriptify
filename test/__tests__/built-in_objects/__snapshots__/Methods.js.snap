// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`method with expression name 1`] = `
"(function(){
 const _={};
 let obj;
 obj={};
 _.a=Symbol(\\"r\\");
 Object.assign(obj,{[_.a](x) {
      return 2 * x;
    }});
 _.b={length:{value:1,configurable:true},name:{value:\\"[r]\\",configurable:true}};
 Object.defineProperties(obj[_.a],_.b);
 return obj;
 })()"
`;

exports[`method with expression name 2`] = `
"(function(){
 const _={};
 let obj;
 obj={f:{}};
 _.a=Symbol(\\"r\\");
 Object.assign(obj.f,{[_.a](x) {
      return 2 * x;
    }});
 _.b={length:{value:1,configurable:true},name:{value:\\"[r]\\",configurable:true}};
 Object.defineProperties(obj.f[_.a],_.b);
 return obj;
 })()"
`;

exports[`method with expression name 3`] = `
"(function(){
 const _={};
 let obj;
 obj={};
 _.b=Symbol(\\"r\\");
 _.a={[_.b]:{set:Object.getOwnPropertyDescriptor({set [\\"set\\"](x) {}},\\"set\\").set,enumerable:true,configurable:true}};
 _.c={name:{value:\\"set [r]\\",configurable:true}};
 Object.defineProperties(_.a[_.b].set,_.c);
 Object.defineProperties(obj,_.a);
 return obj;
 })()"
`;

exports[`method with expression name 4`] = `
"(function(){
 const _={};
 let obj;
 obj={f:{}};
 _.b=Symbol(\\"r\\");
 _.a={[_.b]:{set:Object.getOwnPropertyDescriptor({set [\\"set\\"](x) {}},\\"set\\").set,enumerable:true,configurable:true}};
 _.c={name:{value:\\"set [r]\\",configurable:true}};
 Object.defineProperties(_.a[_.b].set,_.c);
 Object.defineProperties(obj.f,_.a);
 return obj;
 })()"
`;

exports[`method with expression name 5`] = `
"(function(){
 const _={};
 let obj;
 obj={};
 _.b=Symbol(\\"r\\");
 _.a={[_.b]:{get:Object.getOwnPropertyDescriptor({get [\\"get\\"]() {
      return 2;
    }},\\"get\\").get,enumerable:true,configurable:true}};
 _.c={name:{value:\\"get [r]\\",configurable:true}};
 Object.defineProperties(_.a[_.b].get,_.c);
 Object.defineProperties(obj,_.a);
 return obj;
 })()"
`;

exports[`method with expression name 6`] = `
"(function(){
 const _={};
 let obj;
 obj={f:{}};
 _.b=Symbol(\\"r\\");
 _.a={[_.b]:{get:Object.getOwnPropertyDescriptor({get [\\"get\\"]() {
      return 2;
    }},\\"get\\").get,enumerable:true,configurable:true}};
 _.c={name:{value:\\"get [r]\\",configurable:true}};
 Object.defineProperties(_.a[_.b].get,_.c);
 Object.defineProperties(obj.f,_.a);
 return obj;
 })()"
`;
